REGRESSION(303361@main): accessibility/ios-simulator/live-regions/live-region-types.html crashes

hoffmanjoshua opened the pull request:
#### 02c24f670489d015bc73f8f8528bd976d333bd7e
<pre>
REGRESSION(303361@main): accessibility/ios-simulator/live-regions/live-region-types.html crashes
<a href="https://bugs.webkit.org/show_bug.cgi?id=303329">https://bugs.webkit.org/show_bug.cgi?id=303329</a>
<a href="https://rdar.apple.com/165634340">rdar://165634340</a>

Reviewed by Tyler Wilcock.

We were hitting this assert because on iOS, INCLUDE_IGNORED_IN_CORE_AX_TREE is false, and
all child indices are assumed to be based on unignored parents/children.

We can fix this by using unignored parents when traversing using nextInPreOrder, only when
INCLUDE_IGNORED_IN_CORE_AX_TREE is disabled.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::crossFrameParentObjectUnignored const):
(WebCore::AXCoreObject::parentObjectIncludingCrossFrame const):
(WebCore::AXCoreObject::nextInPreOrder):
(WebCore::AXCoreObject::nextSiblingIncludingIgnored const):
* Source/WebCore/accessibility/AXCoreObject.h:

Canonical link: <a href="https://commits.webkit.org/303912@main">https://commits.webkit.org/303912@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/93b135cbf1fa1d9cfe9d1504d904ae099b448ff6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133899 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6410 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45093 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141479 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85959 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135769 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6940 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6274 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102424 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/69732 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136846 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4883 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120032 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83223 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4762 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2382 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113920 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38151 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144122 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6080 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38731 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110792 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6162 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5171 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110998 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28165 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4616 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116290 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59827 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6132 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34558 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5978 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69596 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6223 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6086 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -487,6 +487,20 @@ AXCoreObject* AXCoreObject::parentObjectIncludingCrossFrame() const
+#if ENABLE_ACCESSIBILITY_LOCAL_FRAME
+    if (!result) {
+        if (auto* crossFrameParent = crossFrameParentObject())
+            result = crossFrameParent;
```
I think it would be more efficient to just return `crossFrameParent` here (avoiding a ref-count churn), and return nullptr at the bottom of this function

hoffmanjoshua replied:
Good call! Updated the other method as well while here

twilco commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -479,14 +479,28 @@ AXCoreObject* AXCoreObject::parentObjectIncludingCrossFrame() const
+#endif
+
+    return result.unsafeGet();
+}
```
Can this whole method be written much more simply?

```cpp
if (auto* parent = parentObjectUnignored())
    return parent;
return crossFrameParentObject();
```

Same for `parentObjectIncludingCrossFrame`

twilco replied:
If safer-cpp complains about the `auto* parent`, we can use `SUPPRESS_UNCOUNTED_LOCAL`. Chris Dumez has done this for the same pattern in a different PR.

hoffmanjoshua replied:
Good idea! Fixed up this method, and ended up finding another method that was identical to the one I was adding (`crossFrameParentObjectUnignored`), so updated that one too.

hoffmanjoshua committed:
REGRESSION(303361@main): accessibility/ios-simulator/live-regions/live-region-types.html crashes
https://bugs.webkit.org/show_bug.cgi?id=303329
rdar://165634340

Reviewed by Tyler Wilcock.

We were hitting this assert because on iOS, INCLUDE_IGNORED_IN_CORE_AX_TREE is false, and
all child indices are assumed to be based on unignored parents/children.

We can fix this by using unignored parents when traversing using nextInPreOrder, only when
INCLUDE_IGNORED_IN_CORE_AX_TREE is disabled.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::crossFrameParentObjectUnignored const):
(WebCore::AXCoreObject::parentObjectIncludingCrossFrame const):
(WebCore::AXCoreObject::nextInPreOrder):
(WebCore::AXCoreObject::nextSiblingIncludingIgnored const):
* Source/WebCore/accessibility/AXCoreObject.h:

Canonical link: https://commits.webkit.org/303912@main
